### PR TITLE
Add support for client pagination in GetMetricHistory to the R API

### PR DESCRIPTION
**Pull request by BenWilson2**
Signed-off-by: Ben Wilson <benjamin.wilson@databricks.com>

## Related Issues/PRs

 #xxx

## What changes are proposed in this pull request?

Add support for client-side pagination in `mlflow_get_metric_history` R client API.

## How is this patch tested?

- [X] Existing unit/integration tests
- [ ] New unit/integration tests
- [X] Manual tests (describe details, including test results, below)

CI Shard validation of R API changes:

Screen Shot 2022-12-12 at 5 42 38 PM

## Does this PR change the documentation?

- [ ] No. You can skip the rest of this section.
- [X] Yes. Make sure the changed pages / sections render correctly in the documentation preview.

### Is this a user-facing change?

- [X] No. You can skip the rest of this section.
- [ ] Yes. Give a description of this change to be included in the release notes for MLflow users.

(Details in 1-2 sentences. You can just refer to another PR with a description if this PR is part of a larger change.)

### What component(s), interfaces, languages, and integrations does this PR affect?

Components

- [ ] `area/artifacts`: Artifact stores and artifact logging
- [ ] `area/build`: Build and test infrastructure for MLflow
- [ ] `area/docs`: MLflow documentation pages
- [ ] `area/examples`: Example code
- [ ] `area/model-registry`: Model Registry service, APIs, and the fluent client calls for Model Registry
- [ ] `area/models`: MLmodel format, model serialization/deserialization, flavors
- [ ] `area/recipes`: Recipes, Recipe APIs, Recipe configs, Recipe Templates
- [ ] `area/projects`: MLproject format, project running backends
- [ ] `area/scoring`: MLflow Model server, model deployment tools, Spark UDFs
- [ ] `area/server-infra`: MLflow Tracking server backend
- [X] `area/tracking`: Tracking Service, tracking client APIs, autologging

Interface

- [ ] `area/uiux`: Front-end, user experience, plotting, JavaScript, JavaScript dev server
- [ ] `area/docker`: Docker use across MLflow's components, such as MLflow Projects and MLflow Models
- [ ] `area/sqlalchemy`: Use of SQLAlchemy in the Tracking Service or Model Registry
- [ ] `area/windows`: Windows support

Language

- [X] `language/r`: R APIs and clients
- [ ] `language/java`: Java APIs and clients
- [ ] `language/new`: Proposals for new client languages

Integrations

- [ ] `integrations/azure`: Azure and Azure ML integrations
- [ ] `integrations/sagemaker`: SageMaker integrations
- [ ] `integrations/databricks`: Databricks integrations

<a name="release-note-category"></a>

### How should the PR be classified in the release notes? Choose one:

- [ ] `rn/breaking-change` - The PR will be mentioned in the "Breaking Changes" section
- [ ] `rn/none` - No description will be included. The PR will be mentioned only by the PR number in the "Small Bugfixes and Documentation Updates" section
- [X] `rn/feature` - A new user-facing feature worth mentioning in the release notes
- [ ] `rn/bug-fix` - A user-facing bug fix worth mentioning in the release notes
- [ ] `rn/documentation` - A user-facing documentation change worth mentioning in the release notes

### CODE REVIEW

**Comment by BenWilson2** [comment]
@mlflow-automation autoformat

**Review by dbczumar — APPROVED** [review]
LGTM! Thanks @BenWilson2 ! Before merging, can you confirm that the `timestamp` column of the R dataframe shown in the test case image contains accurate values? All of the timestamps appear to be 0 / beginning of UNIX epoch.

**Comment by dbczumar** [comment]
@BenWilson2 Can we also introduce a test case for get metric history?

**Comment by BenWilson2** [comment]
> LGTM! Thanks @BenWilson2 ! Before merging, can you confirm that the `timestamp` column of the R dataframe shown in the test case image contains accurate values? All of the timestamps appear to be 0 / beginning of UNIX epoch.

I'll add a new timestamp validation where I don't start from Beginning of epoch time and increment in minutes between each write :) I'll admit that validation case looks a bit absurd ;)

**Comment by BenWilson2** [comment]
Screen Shot 2022-12-13 at 4 43 47 PM
Timestamp resolution for buffered entry starttime in validation test

**Comment by BenWilson2** [comment]
Screen Shot 2022-12-13 at 5 06 28 PM

Some less obnoxious (marginally) timestamps showing unique entries for each row and sorted properly with step.

File: mlflow/R/mlflow/R/tracking-runs.R [hunk 1]
```diff
@@ -5,9 +5,9 @@ NULL
 metric_value_to_rest <- function(value) {
   if (is.nan(value)) {
     as.character(NaN)
-  } else if (is.infinite(value) & value > 0) {
+  } else if (is.infinite(value) && value > 0) {
```
Comment (BenWilson2) [inline mlflow/R/mlflow/R/tracking-runs.R:8]: lint warning fix

File: mlflow/R/mlflow/R/tracking-runs.R [hunk 2]
```diff
@@ -309,8 +333,9 @@ mlflow_get_metric_history <- function(metric_key, run_id = NULL, client = NULL)
 #' @template roxlate-client
 #' @param experiment_ids List of string experiment IDs (or a single string experiment ID) to search
 #' over. Attempts to use active experiment if not specified.
-#' @param filter A filter expression over params, metrics, and tags, allowing returning a subset of runs.
-#'   The syntax is a subset of SQL which allows only ANDing together binary operations between a param/metric/tag and a constant.
+#' @param filter A filter expression over params, metrics, and tags, allowing returning a subset
```
Comment (BenWilson2) [inline mlflow/R/mlflow/R/tracking-runs.R:336]: lint warning fix

File: mlflow/R/mlflow/R/tracking-runs.R [hunk 3]
```diff
@@ -482,7 +507,7 @@ mlflow_record_logged_model <- function(model_spec, run_id = NULL, client = NULL)
   c(client, run_id) %<-% resolve_client_and_run_id(client, run_id)
   mlflow_rest("runs", "log-model", client = client, verb = "POST", data = list(
     run_id = run_id,
-    model_json = jsonlite::toJSON(model_spec, auto_unbox=TRUE)
+    model_json = jsonlite::toJSON(model_spec, auto_unbox = TRUE)
```
Comment (BenWilson2) [inline mlflow/R/mlflow/R/tracking-runs.R:510]: lint warning fix

File: mlflow/R/mlflow/R/tracking-runs.R [hunk 4]
```diff
@@ -511,21 +536,26 @@ mlflow_record_logged_model <- function(model_spec, run_id = NULL, client = NULL)
 #' }
 #'
 #' @export
-mlflow_start_run <- function(run_id = NULL, experiment_id = NULL, start_time = NULL, tags = NULL, client = NULL, nested = FALSE) {
+mlflow_start_run <- function(run_id = NULL, experiment_id = NULL,
```
Comment (BenWilson2) [inline mlflow/R/mlflow/R/tracking-runs.R:539]: lint warning fix

File: mlflow/R/mlflow/R/tracking-runs.R [hunk 5]
```diff
@@ -511,21 +536,26 @@ mlflow_record_logged_model <- function(model_spec, run_id = NULL, client = NULL)
...
+                             start_time = NULL, tags = NULL,
+                             client = NULL, nested = FALSE) {
 
   # When `client` is provided, this function acts as a wrapper for `runs/create` and does not register
   #  an active run.
   if (!is.null(client)) {
-    if (!is.null(run_id)) stop("`run_id` should not be specified when `client` is specified.", call. = FALSE)
+    if (!is.null(run_id))
```
Comment (BenWilson2) [inline mlflow/R/mlflow/R/tracking-runs.R:546]: lint warning fix

File: mlflow/R/mlflow/R/tracking-runs.R [hunk 6]
```diff
@@ -511,21 +536,26 @@ mlflow_record_logged_model <- function(model_spec, run_id = NULL, client = NULL)
...
     run <- mlflow_create_run(client = client, start_time = start_time,
                              tags = tags, experiment_id = experiment_id)
     return(run)
   }
 
   # Fluent mode, check to see if extraneous params passed.
 
-  if (!is.null(start_time)) stop("`start_time` should only be specified when `client` is specified.", call. = FALSE)
-  if (!is.null(tags)) stop("`tags` should only be specified when `client` is specified.", call. = FALSE)
+  if (!is.null(start_time))
```
Comment (BenWilson2) [inline mlflow/R/mlflow/R/tracking-runs.R:555]: lint warning fix

File: mlflow/R/mlflow/tests/testthat/test-tracking-runs.R [hunk 7]
```diff
@@ -81,7 +81,7 @@ test_that("mlflow_start_run()/mlflow_end_run() works properly with nested runs",
     if (i > 1) {
       tags <- run$tags[[1]]
       expect_equal(
-        tags[tags$key == "mlflow.parentRunId",]$value,
+        tags[tags$key == "mlflow.parentRunId", ]$value,
```
Comment (BenWilson2) [inline mlflow/R/mlflow/tests/testthat/test-tracking-runs.R:84]: lint warning fix

**Review by WeichenXu123 — APPROVED** [review]
LGTM, thanks!

**Comment by harupy** [comment]
@BenWilson2 What is this binary file? mlflow/R/mlflow/tests/testthat/test/crate.bin

**Comment by BenWilson2** [comment]
> @BenWilson2 What is this binary file? mlflow/R/mlflow/tests/testthat/test/crate.bin

Something that's bin in the repo since the H20 support for R was implemented. https://github.com/mlflow/mlflow/commits/7b8f8b5934560060ef05b1032b42126e0cbeb7d2/mlflow/R/mlflow/tests/testthat/test 
It was updated when I rebuilt the package with the changes to the APIs. I'm not sure if having that up to date is critical for R functionality, but figured that it was part of the build process, it might not be harmful to have it updated.

**Pull request merged**